File: app/connector/delay.py
# Delay Connector
import asyncio
import logging
from typing import Any
from typing import Literal

from pydantic import BaseModel

from .enum import ConnectorType
from app.connector.base import BaseConnector

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class DelayConfig(BaseModel):
    """Configuration for Delay connector."""

    duration: int


class DelayWorkflowStep(BaseModel):
    """definition of a Delay step in a workflow."""

    type: Literal[ConnectorType.DELAY] = ConnectorType.DELAY
    name: str
    config: DelayConfig


class DelayOutput(BaseModel):
    """Output model for Delay connector."""

    type: Literal[ConnectorType.DELAY] = ConnectorType.DELAY
    duration: int
    message: str


class DelayConnector(BaseConnector):
    """Connector that waits for a specified duration."""

    def __init__(self):
        super().__init__(ConnectorType.DELAY)

    async def execute(
        self, step: DelayWorkflowStep, context: dict[str, Any]
    ) -> DelayOutput:
        """
        Wait for specified duration.

        Args:
            step (DelayWorkflowStep): The step configuration.
            context (dict[str, Any]): The execution context.

        Returns:
            DelayOutput: The output containing duration and message.
        """
        assert step.type == ConnectorType.DELAY
        duration = step.config.duration  # Default 1 second

        logger.info(f"Delaying for {duration} seconds")
        await asyncio.sleep(duration)

        return DelayOutput(duration=duration, message=f"Delayed for {duration} seconds")

File: rust-app/src/connector/delay.rs
use async_trait::async_trait;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use tracing::info;

use crate::connector::base::Connector;
use crate::schemas::workflow::{WorkflowStep, WorkflowStepResponse};

/// Configuration for the Delay connector.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DelayConfig {
    pub duration: u64,
}

/// Output model for the Delay connector.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DelayOutput {
    pub duration: u64,
    pub message: String,
}

/// Connector that waits for a specified duration.
pub struct DelayConnector;

impl DelayConnector {
    pub fn new() -> Self {
        Self
    }
}

impl Default for DelayConnector {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl Connector for DelayConnector {
    async fn execute(
        &self,
        step: &WorkflowStep,
        _context: &HashMap<String, serde_json::Value>,
    ) -> Result<WorkflowStepResponse, Box<dyn std::error::Error + Send + Sync>> {
        match step {
            WorkflowStep::Delay { config, .. } => {
                let duration = config.duration;
                info!("Delaying for {} seconds", duration);
                tokio::time::sleep(tokio::time::Duration::from_secs(duration)).await;

                Ok(WorkflowStepResponse::Delay(DelayOutput {
                    duration,
                    message: format!("Delayed for {} seconds", duration),
                }))
            }
            _ => Err("DelayConnector received non-delay step".into()),
        }
    }
}
